_posts/2019-01-13-dask-cudf-first-steps.md

```diff
@@ -0,0 +1,437 @@
+      <td> 170 MB/s </td>
+    </tr>
+    <tr>
+      <th> Forty CPU Cores </th>
+      <td> 35s </td>
+      <td> 285 MB/s </td>
+    </tr>
+    <tr>
+      <th> One GPU </th>
+      <td> 11s </td>
+      <td> 900 MB/s </td>
+    </tr>
+    <tr>
+      <th> Eight GPUs </th>
+      <td> 5s </td>
+      <td> 2000 MB/s </td>
+    </tr>
+  </tbody>
+</table>
+
+*The bandwidth numbers were computed by noting that the data was around 10 GB on disk*
+
+
+Analysis
+--------
+
+First, lets appreciate again that it's easy to test a wide variety of
+architectures using this setup because of the Pandas API compatibility between
+all of the different projects.  We're seeing a wide range of performance (40x
+span) across a variety of different hardware with a wide range of cost points.
+
+Second lets note that this problem scales less well than our
+[previous example with CuPy](../../../2019/01/03/dask-array-gpus-first-steps),
+both on CPU and GPU.
+I suspect that this is because this example is also bound by I/O and not just
+number-crunching.  While the jump from single-CPU to single-GPU is large, the
+jump from single-CPU to many-CPU or single-GPU to many-GPU is not as large as
+we would have liked.  For GPUs for example we got around a 2x speedup when we
+added 8x as many GPUs.
+
+At first one might think that this is because we're saturating disk read speeds.
+However two pieces of evidence go against that guess:
+
+-  NVIDIA folks familiar with my current hardware inform me that they're able to get
+   much more I/O when they're careful (TODO verify)
+-  The CPU scaling is similarly poor, despite the fact that it's obviously not
+   reaching full I/O bandwidth
+
+Instead, it's likely that we're just not treating our disks and IO pipelines
+carefully.  Many simultaneous reads are causing contention, and (TODO need more
+information from people familiar with the DGX).
+
+
+We might consider working to think more carefully about data locality within a
+single machine.  Alternatively, we might just choose to use a smaller machine,
+or many smaller machines.  My team has been asking me to start playing with
+some cheaper systems than a DGX, I may experiment with those soon.  It may be
+that for data-loading and pre-processing workloads the previous wisdom of "pack
+as much computation as you can into a single box" no longer holds
+(without us doing more work that is).
+
+
+### Come help!
+
+If the work above sounds interesting to you then come help!
+There is a lot of low-hanging and high impact work to do.
+
+If you're interested in being paid to focus more on these topics, then consider
+applying for a job.  The NVIDIA corporation is hiring around the use of Dask
+with GPUs.
+
+-  [Senior Library Software Engineer - RAPIDS](https://nvidia.wd5.myworkdayjobs.com/en-US/NVIDIAExternalCareerSite/job/US-TX-Austin/Senior-Library-Software-Engineer---RAPIDS_JR1919608-1)
+
+That's a fairly generic posting.  If you're interested the posting doesn't seem
```

Comment:
Suggest dropping this ending and just say "and other data analytics library development projects" or something similar in the previous sentence.

Reply:
OK, will do.